org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/IReadAccess.java

```diff
@@ -8,22 +8,140 @@
  *******************************************************************************/
 package org.eclipse.xtext.util.concurrent;
 
+import java.util.function.Function;
+import java.util.function.Supplier;
+
 /**
  * @author Sven Efftinge - Initial contribution and API
  * @author Jan Koehnlein - Separated read and write access
  */
-public interface IReadAccess<P> {
+public interface IReadAccess<State> {
+	
+	/**
+	 * Gets a read-only copy of the State and executes {@code work} on it.
+	 * 
+	 * WARNING: the State passed to {@code work} can be null.
+	 * 
+	 * @param work Work to execute on the State
+	 * 
+	 * @return The result of executing {@code work}
+	 * @since 2.7
+	 */
+	<Result> Result readOnly(IUnitOfWork<Result, State> work);
+
+	/**
+	 * Tries to get a read-only copy of the State and execute {@code work} on it.
+	 * 
+	 * @param work Work to execute on the State
+	 * @param defaultResult Supplies a result in case the State is null
+	 * 
+	 * @return The result of executing {@code work}, or
+	 *         the result of querying {@code defaultResult} if the State is null
+	 * @since 2.14
+	 */
+	default <Result> Result tryReadOnly(
+		IUnitOfWork<Result, State> work,
+		Supplier<Result> defaultResult
+	) {
+		return readOnly((state) -> {
+			if (state == null) {
+				return defaultResult.get();
+			}
+			return work.exec(state);
+		});
+	}
 	
-	<T> T readOnly(IUnitOfWork<T,P> work);
+	/**
+	 * Tries to get a read-only copy of the State and execute {@code work} on it.
+	 * 
+	 * @param work Work to execute on the State
+	 * @param defaultResult Supplies a result in case the State is null
+	 * @param exceptionHandler Supplies a result in case an exception is raised during execution
+	 * 
+	 * @return The result of executing {@code work},
+	 *         the result of querying {@code defaultResult} if the State is null, or
+	 *         the result of executing {@code exceptionHandler} in case an exception is raised
+	 * @since 2.14
+	 */
+	default <Result> Result tryReadOnly(
+		IUnitOfWork<Result, State> work,
+		Supplier<Result> defaultResult,
```

Comment:
We could use Supplier<? extends Result> similar to the signature of the exceptionHandler.

Same for the other try's